crates/red_knot_module_resolver/src/resolver.rs

```diff
@@ -143,67 +155,300 @@ impl RawModuleResolutionSettings {
+                    .unwrap(),
+                )
+            })
             .collect();
 
-        paths.push(ModuleResolutionPathBuf::first_party(workspace_root).unwrap());
-
-        paths.push(
-            custom_typeshed.map_or_else(ModuleResolutionPathBuf::vendored_stdlib, |custom| {
-                ModuleResolutionPathBuf::stdlib_from_custom_typeshed_root(&custom).unwrap()
-            }),
-        );
+        static_search_paths.insert(Arc::new(
+            ModuleResolutionPathBuf::first_party(SystemPath::absolute(
+                workspace_root,
+                current_directory,
+            ))
+            .unwrap(),
+        ));
+
+        static_search_paths.insert(Arc::new(custom_typeshed.map_or_else(
+            ModuleResolutionPathBuf::vendored_stdlib,
+            |custom| {
+                ModuleResolutionPathBuf::stdlib_from_custom_typeshed_root(&SystemPath::absolute(
+                    custom,
+                    current_directory,
+                ))
+                .unwrap()
+            },
+        )));
+
+        let mut site_packages = None;
+
+        if let Some(path) = site_packages_setting {
+            let site_packages_root = Arc::new(
+                ModuleResolutionPathBuf::site_packages(SystemPath::absolute(
+                    path,
+                    current_directory,
+                ))
+                .unwrap(),
+            );
+            site_packages = Some(site_packages_root.clone());
+            static_search_paths.insert(site_packages_root);
+        }
 
         // TODO vendor typeshed's third-party stubs as well as the stdlib and fallback to them as a final step
-        if let Some(site_packages) = site_packages {
-            paths.push(ModuleResolutionPathBuf::site_packages(site_packages).unwrap());
-        }
 
         ModuleResolutionSettings {
             target_version,
-            search_paths: OrderedSearchPaths(paths.into_iter().map(Arc::new).collect()),
+            search_path_settings: ValidatedSearchPathSettings {
+                static_search_paths,
+                site_packages,
+            },
+        }
+    }
+}
+
+#[derive(Debug, PartialEq, Eq, Clone)]
+struct ValidatedSearchPathSettings {
+    /// Search paths that have been statically determined purely from reading Ruff's configuration settings.
+    /// These shouldn't ever change unless the config settings themselves change.
+    ///
+    /// We use an `OrderSet` for storing the paths, since the order is important,
+    /// but at runtime, [`sys.path` never contains duplicate entries]
+    ///
+    /// Note that `site-packages` *is included* as a search path in this sequence,
+    /// but it is also stored separately so that we're able to find editable installs later.
+    ///
+    /// [`sys.path` never contains duplicate entries]: https://docs.python.org/3/library/site.html#module-site
+    static_search_paths: OrderSet<SearchPathRoot>,
+    site_packages: Option<SearchPathRoot>,
```

Comment:
No -- we need to know exactly what the site-packages path is later because we only lazily compute the editable-install search paths from the `.pth` files when and if we hit a module resolution that requires it. When we lazily compute the editable-install paths, we need to know the path to `site-packages` so that we can call `read_dir()` on the site-packages path and find all the `.pth` files in it.

Reply:
Sorry. I meant line 240

Reply:
`OrderSet` in this file is currently (following your previous review comments) just defined as

```rs
type OrderSet<T> = indexmap::IndexSet<T, BuildHasherDefault<FxHasher>>;
```

But that's a pretty darn confusing name for me to have given that type alias! So I'll change that.

Reply:
ah that explains it lol.